src/System.Memory/src/System/Buffers/SequenceReader.Search.cs

```diff
@@ -0,0 +1,770 @@
+        /// <param name="delimiterEscape">If found prior to <paramref name="delimiter"/> it will skip that occurrence.</param>
+        /// <param name="advancePastDelimiter">True to move past the <paramref name="delimiter"/> if found.</param>
+        /// <returns>True if the <paramref name="delimiter"/> was found.</returns>
+        public bool TryReadTo(out ReadOnlySequence<T> sequence, T delimiter, T delimiterEscape, bool advancePastDelimiter = true)
+        {
+            SequenceReader<T> copy = this;
+
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            bool priorEscape = false;
+
+            while (_moreData)
+            {
+                int index = remaining.IndexOf(delimiter);
+                if (index != -1)
+                {
+                    if (index == 0 && priorEscape)
+                    {
+                        // We were in the escaped state, so skip this delimiter
+                        priorEscape = false;
+                        Advance(index + 1);
+                        remaining = UnreadSpan;
+                        continue;
+                    }
+                    else if (index > 0 && remaining[index - 1].Equals(delimiterEscape))
+                    {
+                        // This delimiter might be skipped
+
+                        // Count our escapes
+                        int escapeCount = 0;
+                        for (int i = index; i > 0 && remaining[i - 1].Equals(delimiterEscape); i--, escapeCount++)
+                            ;
+                        if (escapeCount == index && priorEscape)
+                        {
+                            // Started and ended with escape, increment once more
+                            escapeCount++;
+                        }
+
+                        priorEscape = false;
+                        if ((escapeCount & 1) != 0)
+                        {
+                            // Odd escape count means we're in the escaped state, so skip this delimiter
+                            Advance(index + 1);
+                            remaining = UnreadSpan;
+                            continue;
+                        }
+                    }
+
+                    // Found the delimiter. Move to it, slice, then move past it.
+                    if (index > 0)
+                    {
+                        Advance(index);
+                    }
+
+                    sequence = Sequence.Slice(copy.Position, Position);
+                    if (advancePastDelimiter)
+                    {
+                        Advance(1);
+                    }
+                    return true;
+                }
+
+                // No delimiter, need to check the end of the span for odd number of escapes then advance
+                {
+                    int escapeCount = 0;
+                    for (int i = remaining.Length; i > 0 && remaining[i - 1].Equals(delimiterEscape); i--, escapeCount++)
+                        ;
```

Comment:
Nit: there seems to be some inconsistency as to whether loops with empty bodies just use `;` or have empty braces.

Reply:
Is this something we've called out in our guidelines? I'll massage it either way.